vms/shift/views.py

```diff
@@ -1,728 +1,751 @@
+    template_name = 'shift/edit_hours.html'
+    form_class = HoursForm
+
+    @method_decorator(check_correct_volunteer)
+    def dispatch(self, *args, **kwargs):
+        return super(EditHoursView, self).dispatch(*args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        context = super(EditHoursView, self).get_context_data(**kwargs)
+        volunteer_id = self.kwargs['volunteer_id']
+        shift_id = self.kwargs['shift_id']
+        context['volunteer_shift'] = get_volunteer_shift_by_id(
+            volunteer_id, shift_id)
+        context['shift'] = get_shift_by_id(shift_id)
+        return context
+
+    def form_valid(self, form):
+        volunteer_id = self.kwargs['volunteer_id']
+        shift_id = self.kwargs['shift_id']
+        shift = get_shift_by_id(shift_id)
+        volunteer_shift = get_volunteer_shift_by_id(volunteer_id, shift_id)
+        start_time = form.cleaned_data['start_time']
+        end_time = form.cleaned_data['end_time']
+        shift_start_time = shift.start_time
+        shift_end_time = shift.end_time
+        try:
+            if (end_time > start_time):
+                if (start_time >= shift_start_time
+                        and end_time <= shift_end_time):
+                    edit_shift_hours(volunteer_id, shift_id, start_time,
+                                     end_time)
+                    return HttpResponseRedirect(
+                        reverse('shift:view_hours', args=(volunteer_id, )))
+                else:
+                    messages.add_message(
+                        self.request, messages.INFO,
+                        'Logged hours should be between shift hours')
+                    return render(
+                        self.request, 'shift/edit_hours.html', {
+                            'form': form,
+                            'shift_id': shift_id,
+                            'volunteer_id': volunteer_id,
+                            'shift': shift,
+                            'volunteer_shift': volunteer_shift,
+                        })
+
+            else:
+                messages.add_message(
+                    self.request, messages.INFO,
+                    'End time should be greater than start time')
+                return render(
+                    self.request, 'shift/edit_hours.html', {
+                        'form': form,
+                        'shift_id': shift_id,
+                        'volunteer_id': volunteer_id,
+                        'shift': shift,
+                        'volunteer_shift': volunteer_shift,
+                    })
+        except:
+            raise Http404
+
+def edit_request(request, shift_id, volunteer_id):
+    vol = get_volunteer_by_id(volunteer_id)
+    shift = get_shift_by_id(shift_id)
+    message = render_to_string('shift/request_edit.txt', {
+                               'volunteer_first_name': vol.first_name,
+                               'volunteer_last_name': vol.last_name,

```

Comment:
Please add a link in the message too that will direct the admin to the approve/cancel request page.